### PR TITLE
[2D] Let's shard every matmul

### PR DESCRIPTION
**Pull request by alanwaketan**
Summary:
This pull requst tries to shard every matmul in LLaMA, below is the sharding strategy:
1. up_proj (batch, length, intermediate): mesh (data, None, model)
2. gate_proj (batch, length, intermediate): mesh (data, None, model)
3. down_proj (batch, length, hidden): mesh (data, None, model)
4. query_states (batch, length, hidden): mesh (data, None, model)
5. key_states (batch, length, hidden / attention_heads * key_value_heads): mesh (data, None, model)
6. value_states (batch, length, hidden / attention_heads * key_value_heads): mesh (data, None, model)
7. attn_weights (batch, num_attention_heads, length, length): mesh (data, model, none, none)
8. attn_output (batch, length, hidden): mesh (data, None, model)
9. hidden_states (batch, length, hidden): mesh (data, None, model)

Test Plan:
Tested on v4-8

### CODE REVIEW

**Comment by alanwaketan** [comment]
I need to rebase. one second...

File: examples/pytorch/language-modeling/run_clm.py [hunk 1]
```diff
@@ -501,76 +506,80 @@ def main():
...
-          import torch_xla
-          print(torch_xla._XLAC._get_xla_sharding_spec(param))
+        # Create an tensor based on the meta tensor
+        # sharded_state_dict[name] = torch.randn_like(param, device=xm.xla_device())
+        # param = sharded_state_dict[name]
+        # # TODO(jonbolin): Can't load_state_dict when the module consists of meta tensors
+        # path = re.sub(r'.(\d+)', r'[\1]', name)
+        # assign = f'model.{path} = torch.nn.Parameter(param)'
+        # # print(f'running "{assign}"')
+        # exec(assign)
```
Comment (JackCaoG) [inline examples/pytorch/language-modeling/run_clm.py]: is this part not needed anymore?

Comment (alanwaketan) [reply examples/pytorch/language-modeling/run_clm.py]: There was a bug in defer_init.

**Comment by alanwaketan** [comment]
After rebasing, the memory increases...
Let me double check...

File: src/transformers/models/llama/modeling_llama.py [hunk 2]
```diff
@@ -686,24 +750,15 @@ def forward(
...
-            data = num_devices // model
-            assert model * data == num_devices
-            data_model_mesh = xs.HybridMesh(ici_mesh_shape=(data, 1, model))
-            xs.mark_sharding(hidden_states, data_model_mesh, (0, 1, 2))
-            if self.spmd_debug:
-                print(torch_xla._XLAC._get_xla_sharding_spec(hidden_states))
+        # Apply 2D sharding:
+        # hidden_states (batch, length, hidden)
+        # mesh (data, None, model)
+        data_model_mesh = xs.HybridMesh(ici_mesh_shape=(self.spmd_data_axis, 1, self.spmd_model_axis))
```
Comment (JackCaoG) [inline src/transformers/models/llama/modeling_llama.py]: I thought we only going to use hybridmesh for v4?

**Review by JackCaoG — APPROVED** [review]
LGTM, thanks!

File: src/transformers/models/llama/modeling_llama.py [hunk 3]
```diff
@@ -314,15 +360,34 @@ def forward(
...
         value_states = value_states.view(bsz, q_len, self.num_key_value_heads, self.head_dim).transpose(1, 2)
 
         kv_seq_len = key_states.shape[-2]
         if past_key_value is not None:
             kv_seq_len += past_key_value[0].shape[-2]
-        cos, sin = self.rotary_emb(value_states, seq_len=kv_seq_len)
-        query_states, key_states = apply_rotary_pos_emb(query_states, key_states, cos, sin, position_ids)
+        # TODO: Disable rotary embedding for now.
+        # cos, sin = self.rotary_emb(value_states, seq_len=kv_seq_len)
+        # query_states, key_states = apply_rotary_pos_emb(query_states, key_states, cos, sin, position_ids)
```
Comment (jonb377) [inline src/transformers/models/llama/modeling_llama.py:403]: Is this needed for 2d sharding?

**Comment by alanwaketan** [comment]
Done rebasing.

**Review by jonb377 — APPROVED** [review]
Thanks Jiewen!

File: src/transformers/models/llama/modeling_llama.py [hunk 4]
```diff
@@ -692,27 +754,18 @@ def forward(
...
-            else:
-                mesh = xs.HybridMesh(ici_mesh_shape=(data, 1, model))
-            xs.mark_sharding(hidden_states, mesh, (0, 1, 2))
-            if self.spmd_debug:
-                print(torch_xla._XLAC._get_xla_sharding_spec(hidden_states))
+        # Apply 2D sharding:
+        # hidden_states (batch, length, hidden)
+        # mesh (data, None, model)
+        if self.spmd_iota_mesh:
+            data_model_mesh = xs.Mesh(device_ids, (self.spmd_data_axis, 1, self.spmd_model_axis))
```
Comment (jonb377) [inline src/transformers/models/llama/modeling_llama.py]: `NameError: name 'device_ids' is not defined`

File: src/transformers/models/llama/modeling_llama.py (hunk 3 above)
Comment (alanwaketan) [reply src/transformers/models/llama/modeling_llama.py:403]: I don't know what's the best way to shard rotary embedding. Just disabled for now.

File: src/transformers/models/llama/modeling_llama.py (hunk 4 above)
Comment (alanwaketan) [reply src/transformers/models/llama/modeling_llama.py]: Good catch. I'm trying to integrate spmd_iota_mesh now.

File: src/transformers/models/llama/modeling_llama.py [hunk 5]
```diff
@@ -187,6 +194,25 @@ def apply_rotary_pos_emb(q, k, cos, sin, position_ids):
...
+    assert model.spmd_data_axis * model.spmd_model_axis == num_devices
+    model.spmd_iota_mesh = config.spmd_iota_mesh
+
+def get_mesh(spmd_iota_mesh, ici_mesh_shape, dcn_mesh_shape=None):
+    if spmd_iota_mesh:
+        if dcn_mesh_shape is not None:
+            assert len(ici_mesh_shape) == len(dcn_mesh_shape)
+            for i in range(len(dcn_mesh_shape)):
+                ici_mesh_shape[i] *= dcn_mesh_shape[i]
+        device_ids = torch.arange(num_devices)
```
Comment (jonb377) [inline src/transformers/models/llama/modeling_llama.py:213]: num_devices isn't defined here

Comment (alanwaketan) [reply src/transformers/models/llama/modeling_llama.py:213]: Oops.

**Comment by alanwaketan** [comment]
@jonb377  @JackCaoG Should I land the change now? Thanks for the review!

**Comment by JackCaoG** [comment]
I am fine with it

**Pull request merged**